nixos/modules/services/hardware/nbfc-linux.nix

```
@@ -0,0 +1,61 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+with lib;
+
+let
+  cfg = config.services.nbfc-linux;
+
+  nbfcConfigList = splitString "\n" (
+    removeSuffix "\n" (
+      builtins.readFile (
+        pkgs.runCommand "nbfc-config-list" { } ''
+          ${pkgs.nbfc-linux}/bin/nbfc config --list > $out
+        ''
+      )
+    )
+  );
+in
+{
+  options.services.nbfc-linux = {
+    enable = lib.mkEnableOption "NBFC: NoteBook FanControl service";
+
+    package = lib.mkPackageOption pkgs "nbfc-linux" { default = "nbfc-linux "; };
+
+    configName = mkOption {
+      type = types.nullOr types.str;
+      description = "The NBFC configuration name to use. Obtain the list via `nbfc config --list`";
+      default = null;
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    assertions = [
+      {
+        assertion = (cfg.configName == null) || (builtins.elem cfg.configName nbfcConfigList);
+        message = "nbfc-linux: invalid configName: ${cfg.configName}";
+      }
+    ];
+
+    environment.systemPackages = [ pkgs.nbfc-linux ];
+
+    environment.etc."nbfc-linux/config.json".source = pkgs.writeTextFile {
+      name = "nbfc-linux-config.json";
+      text = ''{"SelectedConfigId": "${cfg.configName}"}'';
+    };
+
+    systemd.services.nbfc-linux = {
+      enable = true;
+      description = "NoteBook FanControl";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig.Type = "simple";
+      path = [ pkgs.kmod ];
+      script = "${pkgs.nbfc-linux}/bin/nbfc_service --config-file /etc/nbfc-linux/config.json";
+    };
```

Pass the config directly.
```suggestion
    systemd.services.nbfc-linux = {
      description = "NoteBook FanControl";
      wantedBy = [ "multi-user.target" ];
      serviceConfig = {
        Type = "simple";
        ExecStart = "${lib.getExe pkgs.nbfc-linux} --config-file ${pkgs.writeText "nbfc-config.json" (lib.strings.toJSON { SelectedConfigId = cfg.configName; })}";
      };
      path = [ pkgs.kmod ];
    };
```

---

done

---

upstreams systemd unit (we should probably use that one instead also) uses `nbfc start` and `nbfc stop`, why do we need to differ here?

---

we probably don't need to differ, but might have to go back to `environment.etc."nbfc-linux/config.json".source ...` stuff because the `nbfc` exe doesn't accept a `--config-file` flag. or would multiple commands in `ExecStartPre` be better?

e.g.
```ExecStartPre = "${lib.getExe pkgs.nbfc-linux} config --set '${cfg.configName}' && ${lib.getExe pkgs.nbfc-linux} wait-for-hwmon"```

---

i have the nbfc-linux package installed and its nbfc_service.service doesn't seem to be listed on my system so not sure how to leverage it.

---

oh when `ExecStart = "${lib.getExe pkgs.nbfc-linux} start";` I get an error in the service output: `nbfc: ERROR: Can't run nbfc_service, make sure the binary is installed` :confounded:

---

aight so... now not passing the config directly but it is using the `nbfc` exe like upstream. thoughts?